### ISSUE TITLE
Change curve type fields in ThermalStorage:Ice:Detailed

### ISSUE THREAD

**Issue by nealkruis** [NREL/EnergyPlus#6879]
Issue overview
--------------
While removing the curve type requirements in #6810, I noticed an undocumented change in meaning between two choices of 2D curve types. The keys for these fields could something like "ChargeFractionLMTD"/"LMTDMassFlow" and "DischargeFractionLMTD"/"LMTDMassFlow". There's precedent for this elsewhere like Chiller:Electric:ReformulatedEIR (Electric Input to Cooling Output Ratio Function of Part Load Ratio Curve Type = LeavingCondenserWaterTemperature or Lift), and in Boilers with Entering/Leaving water temperature. This should also remove the need for a string comparison on curve object type.

The I/O Ref and Engr. Ref should document both options (currently, only one is mentioned).

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by RKStrand** [comment]
@nealkruis @mjwitte Just starting to look at this, but I want to clarify what is being requested.  So, I will state what I think is being asked and would appreciate it if one or both of you would respond so that I am on the right page.

What I understand is that the problem here is sort of two fold.  First, there are only two potential curve types that are allowed here and there is a desire to open this to ANY curve type involving two variables.  Second, the two curves that are currently allowed have two different variables (and the common one between the two show up in different orders).

So, what is being asked here is that this get switched from two specific curve types to two different sets of independent variables that could use any curve type that has two independent variables.  As part of this, modify the documentation to reflect this change and make sure that the curve descriptions and ice storage input descriptions are complete.  Is that correct?

**Comment by mjwitte** [comment]
Correct.

**Comment by RKStrand** [comment]
@nealkruis @mjwitte I have also uncovered an error in the Detailed Ice Storage Model.  In line 641 of IceThermalStorage.cc:

if (DetIceStor(IceNum).DischargeCurveTypeNum == CurveQuadraticLinear) {

But this is in charging mode, so this should say:

if (DetIceStor(IceNum).ChargeCurveTypeNum == CurveQuadraticLinear) {

Also need to modify the comment two lines after this.  My fix for this CR will correct this problem.

**Comment by RKStrand** [comment]
@nealkruis @mjwitte As I look at this longer, I am wondering if this is sort of turning into almost a small development task rather than a simple fix so this is kind of like the design documentation.  Let's say that we modify the current input which asks for discharging curve object type, discharging curve object name, charging curve object type, and charging curve object name to discharging variable(s), discharging curve object name, charging variables(s), and charging curve object name.  Now, let's say that for charging and discharging variables we have the following possibilities: fraction charged/discharged, LMTD, and mass flow rate.  So, we could have a variety of options: only one variable (3 options), combinations of two variables (3 options...or is it actually six because what about putting the variables in reverse order), or all three variables (1 option).  This gives us a total of 7 or 10 different combinations, some using a single variable, some with two variables, and one with all three.  These would have to be checked to make sure that the curve they are trying to access have the correct number of variables.  Here are my suggestions as far as inputs for the independent variable combinations in the input:

Fraction (fraction charged or discharged)
LMTD (normalized log mean temperature difference)
Flow (normalized mass flow rate)
FractionLMTD (fraction as the first variable, LMTD the second)
FractionFlow  (fraction as the first variable, mass flow the second)
LMTDFlow (LMTD as the first variable, mass flow the second)
FlowLMTD (mass flow as the first variable, LMTD the second)
FlowFraction (mass flow as the first variable, fraction the second)
LMTDFraction (LMTD as the first variable, fraction the second)
FractionLMTDFlow fraction as the first variable, LMTD the second, mass flow as the third)

The curve chosen would be checked against this option to make sure that the number of variables matches.  Any valid curve should be able to be used.  What do you think of this implementation scheme?  Good?  Too much?

**Comment by mjwitte** [comment]
@RKStrand No, we don't need to allow for different orders of the variables (we don't do that anywhere else). 
The intent here was just to support and document clearly the two rather different methods that are supported in the current code for [discharging](https://github.com/NREL/EnergyPlus/blob/7a7082a357ad875ac3f0d9dd66564a85cae454d2/src/EnergyPlus/IceThermalStorage.cc#L641-L645) and [charging](https://github.com/NREL/EnergyPlus/blob/7a7082a357ad875ac3f0d9dd66564a85cae454d2/src/EnergyPlus/IceThermalStorage.cc#L612-L616) which are controlled by the curve types.

**Comment by RKStrand** [comment]
@mjwitte So, the fact that one equation type uses Fraction Charged as the first variable and LMTDstar as the second variable and the other uses LMTDstar as the first variable and Mdotnorm as the second isn't a problem (that LMTDstar show up first in one and second in the other)?  As long as it is explained and it can use any valid curve form (not just the two current set in the code), then that will satisfy the concerns here?

**Comment by mjwitte** [comment]
Correct.

**Comment by RKStrand** [comment]
@mjwitte Got it--thanks!  I will proceed on this basis.

**Comment by mjwitte** [comment]
And, there's a couple of places that use (1-AvgFracCharged)?
```
 Qstar = std::abs(CurveValue(DetIceStor(IceNum).DischargeCurveNum, (1.0 - AvgFracCharged), LMTDstar));
```

**Comment by RKStrand** [comment]
@mjwitte @nealkruis Yes, those are for discharging and use percent discharged.  I'm going to make that clearer by having these options for discharging:

FractionDischargedLMTD
LMTDMassFlow

for the fifth character string in the input (what used to be Discharging Curve Object Type but will now be Discharging Curve Independent Variables)

For charging, the options will be:

FractionChargedLMTD
LMTDMassFlow

for the seventh character string in the input (what used to be Charging Curve Object Type but will now be Charging Curve Independent Variables).

I will explain that fraction discharged is 1-FractionCharged in the documentation edits/additions.

So, this will allow the user to use any curve (though I will check to see that it has two independent variables--users can still do a single dependent variable by using curves that have some coefficients set to zero).

**Comment by mjwitte** [comment]
OK, but I see now why I was confused about fraction charged vs fraction discharged. [This code](https://github.com/NREL/EnergyPlus/blob/7a7082a357ad875ac3f0d9dd66564a85cae454d2/src/EnergyPlus/IceThermalStorage.cc#L641-L646) is checking `DischargeCurveTypeNum` but it's evaluating the ChargeCurve - looks like this should be checking `ChargeCurveTypeNum` here.

**Comment by RKStrand** [comment]
Yes, I agree--see my comment above (third comment after initial description of the defect).  I uncovered that while looking through the code as I started to investigate this and I will correct this as part of this work.

Sorry for any confusion this error caused!

**Comment by mjwitte** [comment]
Closed via #7339
